File: autozig-math/zig/ray.zig
const std = @import("std");

pub const Ray2d = extern struct {
    origin: Vec2,
    direction: Dir2,

    pub fn new(origin: Vec2, direction: Dir2) Ray2d {
        return .{
            .origin = origin,
            .direction = direction,
        };
    }

    pub fn get_point(self: Ray2d, distance: f32) Vec2 {
        const dir_vec = self.direction.vec;
        return .{
            .x = self.origin.x + dir_vec.x * distance,
            .y = self.origin.y + dir_vec.y * distance,
        };
    }

    pub fn intersect_plane(self: Ray2d, plane_origin: Vec2, plane: Plane2d) f32 {
        const denominator = plane.normal.vec.dot(self.direction.vec);
        if (@abs(denominator) > std.math.floatEps(f32)) {
            const diff: Vec2 = .{ .x = plane_origin.x - self.origin.x, .y = plane_origin.y - self.origin.y };
            const distance = diff.dot(plane.normal.vec) / denominator;
            if (distance > std.math.floatEps(f32)) {
                return distance;
            }
        }
        return std.math.nan(f32);
    }
};

pub const Ray3d = extern struct {
    origin: Vec3,
    direction: Dir3,

    pub fn new(origin: Vec3, direction: Dir3) Ray3d {
        return .{
            .origin = origin,
            .direction = direction,
        };
    }

    pub fn get_point(self: Ray3d, distance: f32) Vec3 {
        const dir_vec = self.direction.vec;
        return .{
            .x = self.origin.x + dir_vec.x * distance,
            .y = self.origin.y + dir_vec.y * distance,
            .z = self.origin.z + dir_vec.z * distance,
        };
    }

    pub fn intersect_plane(self: Ray3d, plane_origin: Vec3, plane: InfinitePlane3d) f32 {
        const denominator = plane.normal.vec.dot(self.direction.vec);
        if (@abs(denominator) > std.math.floatEps(f32)) {
            const diff: Vec3 = .{ .x = plane_origin.x - self.origin.x, .y = plane_origin.y - self.origin.y, .z = plane_origin.z - self.origin.z };
            const distance = diff.dot(plane.normal.vec) / denominator;
            if (distance > std.math.floatEps(f32)) {
                return distance;
            }
        }
        return std.math.nan(f32);
    }
};

export fn ray2d_new(origin: Vec2, direction: Dir2) Ray2d {
    return Ray2d.new(origin, direction);
}

export fn ray2d_get_point(self: Ray2d, distance: f32) Vec2 {
    return self.get_point(distance);
}

export fn ray2d_intersect_plane(self: Ray2d, plane_origin: Vec2, plane: Plane2d) f32 {
    return self.intersect_plane(plane_origin, plane);
}

export fn ray3d_new(origin: Vec3, direction: Dir3) Ray3d {
    return Ray3d.new(origin, direction);
}

export fn ray3d_get_point(self: Ray3d, distance: f32) Vec3 {
    return self.get_point(distance);
}

export fn ray3d_intersect_plane(self: Ray3d, plane_origin: Vec3, plane: InfinitePlane3d) f32 {
    return self.intersect_plane(plane_origin, plane);
}

File: autozig-math/src/ray.rs
use autozig::include_zig;
use crate::{Vec2, Vec3, Dir2, Dir3, Plane2d, InfinitePlane3d};

#[repr(C)]
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Ray2d {
    pub origin: Vec2,
    pub direction: Dir2,
}

#[repr(C)]
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Ray3d {
    pub origin: Vec3,
    pub direction: Dir3,
}

include_zig!("zig/ray.zig", {
    fn ray2d_new(origin: Vec2, direction: Dir2) -> Ray2d;
    fn ray2d_get_point(self_: Ray2d, distance: f32) -> Vec2;
    fn ray2d_intersect_plane(self_: Ray2d, plane_origin: Vec2, plane: Plane2d) -> f32;

    fn ray3d_new(origin: Vec3, direction: Dir3) -> Ray3d;
    fn ray3d_get_point(self_: Ray3d, distance: f32) -> Vec3;
    fn ray3d_intersect_plane(self_: Ray3d, plane_origin: Vec3, plane: InfinitePlane3d) -> f32;
});

impl Ray2d {
    pub fn new(origin: Vec2, direction: Dir2) -> Self { ray2d_new(origin, direction) }
    pub fn get_point(&self, distance: f32) -> Vec2 { ray2d_get_point(*self, distance) }
    pub fn intersect_plane(&self, plane_origin: Vec2, plane: Plane2d) -> Option<f32> {
        let dist = ray2d_intersect_plane(*self, plane_origin, plane);
        if dist.is_nan() { None } else { Some(dist) }
    }
}

impl Ray3d {
    pub fn new(origin: Vec3, direction: Dir3) -> Self { ray3d_new(origin, direction) }
    pub fn get_point(&self, distance: f32) -> Vec3 { ray3d_get_point(*self, distance) }
    pub fn intersect_plane(&self, plane_origin: Vec3, plane: InfinitePlane3d) -> Option<f32> {
        let dist = ray3d_intersect_plane(*self, plane_origin, plane);
        if dist.is_nan() { None } else { Some(dist) }
    }
}
